flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/EventTimeTemporalJoinRewriteRule.java

```diff
@@ -0,0 +1,477 @@
+            assert isEventTime(snapshot.getPeriod().getType());
+            final RelNode child = snapshot.getInput();
+            final RelNode newChild = transmitSnapshotRequirement(child);
+            if (newChild != child) {
+                return snapshot.copy(snapshot.getTraitSet(), newChild, snapshot.getPeriod());
+            }
+            return snapshot;
+        }
+        if (node instanceof HepRelVertex) {
+            return transmitSnapshotRequirement(((HepRelVertex) node).getCurrentRel());
+        }
+        if (node instanceof FlinkLogicalWatermarkAssigner) {
+            final FlinkLogicalWatermarkAssigner wma = (FlinkLogicalWatermarkAssigner) node;
+            final RelNode child = wma.getInput();
+            final RelNode newChild = transmitSnapshotRequirement(child);
+            if (newChild != child) {
+                return wma.copy(
+                        wma.getTraitSet(), newChild, wma.rowtimeFieldIndex(), wma.watermarkExpr());
+            }
+            return wma;
+        }
+        if (node instanceof FlinkLogicalTableSourceScan) {
+            final FlinkLogicalTableSourceScan ts = (FlinkLogicalTableSourceScan) node;
+            // update eventTimeSnapshotRequired to true
+            return ts.copy(ts.getTraitSet(), ts.relOptTable(), true);
+        }
+        return node;
+    }
+
+    private boolean isEventTime(RelDataType period) {
+        if (period instanceof TimeIndicatorRelDataType) {
+            return ((TimeIndicatorRelDataType) period).isEventTime();
+        }
+        return false;
+    }
+
+    /**
+     * Configuration for {@link EventTimeTemporalJoinRewriteRule}.
+     *
+     * <p>Operator tree:
+     *
+     * <pre>{@code
+     *    Join (event time temporal)
+     *      /       \
+     * RelNode     [Calc]
+     *               \
+     *             Snapshot
+     *                \
+     *              [Calc]
+     *                 \
+     *             WatermarkAssigner
+     *                  \
+     *                [Calc]
+     *                   \
+     *                TableScan
+     * }</pre>
+     *
+     * <p>8 variants:
+     *
+     * <ul>
+     *   <li>JOIN_CALC_SNAPSHOT_CALC_WMA_CALC_TS
+     *   <li>JOIN_CALC_SNAPSHOT_CALC_WMA_TS
+     *   <li>JOIN_CALC_SNAPSHOT_WMA_CALC_TS
+     *   <li>JOIN_CALC_SNAPSHOT_WMA_TS
+     *   <li>JOIN_SNAPSHOT_CALC_WMA_CALC_TS
+     *   <li>JOIN_SNAPSHOT_CALC_WMA_TS
+     *   <li>JOIN_SNAPSHOT_WMA_CALC_TS
+     *   <li>JOIN_SNAPSHOT_WMA_TS
+     * </ul>
+     *
+     * <p>Note: the binary form of the numeric suffix is consistent with the arrangement of the
+     * three optional calc nodes, e.g., '5' corresponds to '101' in binary, representing the
+     * existence of the 1st and 3rd calc nodes.
+     */
+    public interface Config extends RelRule.Config {
+        RelRule.Config JOIN_CALC_SNAPSHOT_CALC_WMA_CALC_TS =
+                EMPTY.withDescription("EventTimeTemporalJoinRewriteRule7")
```

Comment:
just one minor comment:

how about we just add the unique part of field name as the postfix of the description? just like EventTimeTemporalJoinRewriteRule_CALC_SNAPSHOT_CALC_WMA_CALC

it’s easier to maintain ?

Reply:
make sense to me, updated